### PR TITLE
Support multiple simultaneous mDNS requests

### LINKED ISSUE #21914: net: dns: Answers to multiple mDNS queries sent in parallel aren't properly handled
When trying to resolve several different mDNS names at the same time, the answers received for each queries aren't always properly matched to the initial query.

This is because mDNS queries always have an ID of 0, so when receiving an answer in dns_read, the query slot returned will always be the first of the list since all IDs are at 0:
https://github.com/zephyrproject-rtos/zephyr/blob/404aaec44dca5c2849c4e5308df30fcd7c28be99/subsys/net/lib/dns/resolve.c#L396

RFC 6762 says:
> In multicast responses, including unsolicited multicast responses, the Query Identifier MUST be set to zero on transmission, and MUST be ignored on reception.

Another way to match the answer to the query must be used for mDNS queries.

### PR DESCRIPTION
**Pull request by jukkar**
Fixes #21914

### CODE REVIEW

File: subsys/net/lib/dns/resolve.c [hunk 1]
```diff
@@ -755,14 +813,73 @@ int dns_resolve_cancel(struct dns_resolve_context *ctx, u16_t dns_id)
...
+		len = strlen(query_name);
+		if ((len + 1 + 2) > buf->size) {
+			net_buf_unref(buf);
+			return -ENOMEM;
+		}
+
+		ret = dns_msg_pack_qname(&len, buf->data, buf->size,
+					 query_name);
+		if (ret >= 0) {
+			len = strlen(buf->data);
```
Comment (tbursztyka) [inline subsys/net/lib/dns/resolve.c]: dns_msg_pack_qname() already sets the len, so no need to recompute it.

That removes again a cpu greedy strlen()

File: subsys/net/lib/dns/resolve.c [hunk 2]
```diff
@@ -755,14 +813,73 @@ int dns_resolve_cancel(struct dns_resolve_context *ctx, u16_t dns_id)
...
+			net_buf_add(buf, len + 1);
```
Comment (tbursztyka) [inline subsys/net/lib/dns/resolve.c]: and it did take into account the '\0' already so no need to add + 1 here.

File: subsys/net/lib/dns/resolve.c [hunk 3]
```diff
@@ -755,14 +813,73 @@ int dns_resolve_cancel(struct dns_resolve_context *ctx, u16_t dns_id)
...
+			net_buf_add_be16(buf, query_type);
+
+			query_hash = crc16_ansi(buf->data, len + 1 + 2);
```
Comment (tbursztyka) [inline subsys/net/lib/dns/resolve.c]: then same: + 1 unnecessary

File: subsys/net/lib/dns/resolve.c [hunk 4]
```diff
@@ -755,14 +813,73 @@ int dns_resolve_cancel(struct dns_resolve_context *ctx, u16_t dns_id)
...
+		}
+
+		/* If the query string + \0 + query type (A or AAAA)
+		 * does not fit the tmp buf, then bail out
+		 */
+		len = strlen(query_name);
+		if ((len + 1 + 2) > buf->size) {
+			net_buf_unref(buf);
+			return -ENOMEM;
+		}
```
Comment (tbursztyka) [inline subsys/net/lib/dns/resolve.c]: this whole test is useless: dns_msg_pack_qname() will already check if the query fits, and later: before the net_buf_add_be16(), just check len + 2 against buf->size then you'll get your -ENOMEM error. 

That removes a strlen() which is greedy.

**Comment by jukkar** [comment]
Updated according to comments.

**Comment by jukkar** [comment]
ping @tbursztyka, can you recheck

**Comment by pfalcon** [comment]
cc @lbise

**Review by pfalcon — COMMENTED** [review]
I don't understand the logic behind introducing hash usage here. Hash is needed to speed up comparisons in large search domain. For example, if you have a million entries, you definitely need a hash. How many in-flight requests we'd have here? A dozen? Then, hash has collisions. In other words, if hashes are equal, it doesn't mean that subject datums are equal, that's the purpose of hash - to quickly vet out inequalities, but if hashes are equal, you need to compare the strings themselves. Finally, here crc16 is used as a hash function. Our implementation is optimized for code size, and is thus slow (works on bit level). If we'd have optimization for runtime, it would ship a precomputed value table, which would be mean code size dependency.

So, that's what I see here: using slow hash function -> using it to compare within just a handful of choices -> while doing so, not handling hash collisions. You could instead compare query host name directly, which would be simpler code without additional dependencies, and would be only faster for the realistic cases we have in hand.

I'm sorry if I missed something here.

**Comment by jukkar** [comment]
> I don't understand the logic behind introducing hash usage here. Hash is needed to speed up comparisons in large search domain. For example, if you have a million entries, you definitely need a hash. How many in-flight requests we'd have here? A dozen? Then, hash has collisions. In other words, if hashes are equal, it doesn't mean that subject datums are equal, that's the purpose of hash - to quickly vet out inequalities, but if hashes are equal, you need to compare the strings themselves. Finally, here crc16 is used as a hash function. Our implementation is optimized for code size, and is thus slow (works on bit level). If we'd have optimization for runtime, it would ship a precomputed value table, which would be mean code size dependency.

You missed the problem domain that this PR is trying to solve. So the patch tries to solve the issue of matching sent mDNS request to received reply. Previously we used DNS id for matching the reply but it cannot be used for that purpose in mDNS as the id is 0. So this PR calculates a "hash" which is a crc16 value in this case. I used crc16 as I did not want to re-invent a hash function for this and there was simple API for calculating the crc value. The 16-bit value was used in order not to bloat the dns_resolve_context struct too much. The mDNS resolving is not done many times in a row so speed is not an issue here. Currently crc functions are optimized for code size, when / if we get hw accelerated crc functions there is no possible speed issue.

> 
> So, that's what I see here: using slow hash function -> using it to compare within just a handful of choices -> while doing so, not handling hash collisions. You could instead compare query host name directly, which would be simpler code without additional dependencies, and would be only faster for the realistic cases we have in hand.

That was one alternative, it is just that it would bloat things considerably in dns_resolve_context struct. We store pointer to sent query, but it is just a normal string. When receiving the mDNS response, the response is in DNS string format and we cannot easily compare these. We would need to allocate space somewhere etc. which would mean bigger memory consumption.

**Review by tbursztyka — APPROVED** [review]

**Comment by pfalcon** [comment]
@jukkar 

> You missed the problem domain that this PR is trying to solve. So the patch tries to solve the issue of matching sent mDNS request to received reply.

Yes, and that's done by matching reply against array of outstanding requests, i.e. searching within the array of outstanding requests, so where's my missing of the problem domain?

> I used crc16 as I did not want to re-invent a hash function for this and there was simple API for calculating the crc value.

That's understood, but my comment is that CRC is known not to be an ideal hash function. If you go as far as bringing in a hash function at all, could add one of well-known both-simple-and-fast functions, e.g. "DJB hash": http://www.cse.yorku.ca/~oz/hash.html .

> We store pointer to sent query, but it is just a normal string. When receiving the mDNS response, the response is in DNS string format and we cannot easily compare these. We would need to allocate space somewhere etc. which would mean bigger memory consumption.

My point still holds: hash function values collide, so when you get hash match, next step always should be comparing literal (unhashed) content. I don't see that being done, get_slot_by_id()  https://github.com/zephyrproject-rtos/zephyr/pull/21952/files#diff-31ea32ed325f4d523a7f77df9b060689R345 is just extended to check hash value (perhaps its name should be changed then?), and it just stops on this match returning the slot index, and as far as I can tell, this return value is used as-is later.

So, you would need to analyze what may happen in the case of hash collision with your solution. I'd say, this opens up a security issue, when request is made to resolve e.g. "foo.com", but reply is used for e.g. "bar.com". This could be done on purpose (collisions for CRC aren't hard to find), and could be used to redirect traffic for one site to another site.

**Comment by jukkar** [comment]
> hash function values collide

Indeed they do. I verified that we do not collide with IPv4 and IPv6 mDNS and DNS queries that we send at the same time (can be checked by samples/net/dns_resolve sample app). This is the most important test here, it is very unlikely that the application would do a query at the same time that would collide with currently pending DNS query. Hash collision is of course possible but IMHO for all the practical purposes the current code is good enough.

**Comment by lbise** [comment]
I'll test the changes

**Comment by lbise** [comment]
Tested looks good. Thank you

**Pull request merged**